### PR TITLE
feat: Implements `mongodbatlas_stream_privatelink_endpoint` data sources

### PR DESCRIPTION
**Pull request by oarbusi**
## Description

Implements `mongodbatlas_stream_privatelink_endpoint` data sources

Example usage: 
```
data "mongodbatlas_stream_privatelink_endpoint" "singular-datasource-test" {
  project_id = var.project_id
  id         = <id>
}

data "mongodbatlas_stream_privatelink_endpoints" "plural-datasource-test" {
  project_id = var.project_id
}
```

Link to any related issue(s): CLOUDP-289663

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [ ] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals I have added appropriate changelog entries.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/service/streamprivatelinkendpoint/model.go [hunk 1]
```diff
@@ -23,9 +22,9 @@ func NewTFModel(ctx context.Context, projectID string, apiResp *admin.StreamsPri
 		Vendor:              types.StringPointerValue(apiResp.Vendor),
 	}
 	if apiResp.DnsSubDomain != nil {
-		subdomain, diag := types.ListValueFrom(ctx, types.StringType, apiResp.GetDnsSubDomain())
-		if diag.HasError() {
-			return nil, diag
+		subdomain, diagn := types.ListValueFrom(ctx, types.StringType, apiResp.GetDnsSubDomain())
```
Comment (EspenAlbert) [inline internal/service/streamprivatelinkendpoint/model.go]: [nit] In recently generated model conversions we prefer sending `diags` as a parameter and simply return `*TFModel`, that simplifies the logic a big.
I think `diags` is a better name than `diagn`

File: internal/service/streamprivatelinkendpoint/plural_data_source.go [hunk 2]
```diff
@@ -38,19 +41,28 @@ func (d *pluralDS) Read(ctx context.Context, req datasource.ReadRequest, resp *d
...
-
-	// TODO: process response into new terraform state
-	// newStreamPrivatelinkEndpointsModel, diags := NewTFModelPluralDS(ctx, apiResp)
-	// if diags.HasError() {
-	// 	resp.Diagnostics.Append(diags...)
-	// 	return
-	// }
-	// resp.Diagnostics.Append(resp.State.Set(ctx, newStreamPrivatelinkEndpointsModel)...)
+	projectID := tfModel.ProjectId.ValueString()
+
```
Comment (EspenAlbert) [inline internal/service/streamprivatelinkendpoint/plural_data_source.go]: Maybe remove some linebreaks?

File: internal/service/streamprivatelinkendpoint/plural_data_source.go [hunk 3]
```diff
@@ -38,19 +41,28 @@ func (d *pluralDS) Read(ctx context.Context, req datasource.ReadRequest, resp *d
...
+	connV2 := d.Client.AtlasV2
+
+	params := admin.ListPrivateLinkConnectionsApiParams{
+		GroupId: projectID,
+	}
+
+	streamsPrivateLinkConnections, err := dsschema.AllPages(ctx, func(ctx context.Context, pageNum int) (dsschema.PaginateResponse[admin.StreamsPrivateLinkConnection], *http.Response, error) {
```
Comment (EspenAlbert) [inline internal/service/streamprivatelinkendpoint/plural_data_source.go:50]: Nice usage of `dsschema.AllPages`!

File: internal/service/streamprivatelinkendpoint/model.go [hunk 4]
```diff
@@ -55,23 +54,21 @@ func NewAtlasReq(ctx context.Context, plan *TFModel) (*admin.StreamsPrivateLinkC
...
-	// }
-	// if diags.HasError() {
-	// 	return nil, *diags
-	// }
-	// return &TFModelDSP{
-	// 	ProjectId: types.StringValue(projectID),
-	// 	Results:   tfModels,
-	// }, *diags
-	return nil, nil
+func NewTFModelPluralDS(ctx context.Context, projectID string, sdkResults []admin.StreamsPrivateLinkConnection) (*TFModelDSP, diag.Diagnostics) {
```
Comment (EspenAlbert) [inline internal/service/streamprivatelinkendpoint/model.go:57]: Same comment as [above](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2897#discussion_r1886920637)

File: templates/data-source.md.tmpl [hunk 5]
```diff
@@ -55,6 +55,7 @@
     {{ else if eq .Name "mongodbatlas_third_party_integration" }}
     {{ else if eq .Name "mongodbatlas_x509_authentication_database_user" }}
     {{ else if eq .Name "mongodbatlas_privatelink_endpoint_service_data_federation_online_archive" }}
+    {{ else if eq .Name "mongodbatlas_stream_privatelink_endpoint" }}
```
Comment (EspenAlbert) [inline templates/data-source.md.tmpl:58]: Are doc updates not done in a different PR?

**Review by EspenAlbert — APPROVED** [review]
LGTM. A few nit comments

File: templates/data-source.md.tmpl (hunk 5 above)
Comment (oarbusi) [reply templates/data-source.md.tmpl:58]: yes it will be done in a different PR, but since I am adding the DS into the provider, without this, the generate-doc-all GH action would fail

File: internal/service/streamprivatelinkendpoint/data_source.go [hunk 6]
```diff
@@ -37,19 +36,20 @@ func (d *ds) Read(ctx context.Context, req datasource.ReadRequest, resp *datasou
...
+		resp.Diagnostics.AddError("error fetching resource", err.Error())
+		return
+	}
+
+	newStreamPrivatelinkEndpointModel, diags := NewTFModel(ctx, projectID, streamsPrivateLinkConnection)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+	resp.Diagnostics.Append(resp.State.Set(ctx, newStreamPrivatelinkEndpointModel)...)
```
Comment (lantoli) [inline internal/service/streamprivatelinkendpoint/data_source.go:54]: similar to @EspenAlbert comment, if you use diags as input param, it's a bit more straightforward and you don't need to assign it back, e.g.: 

```suggestion
	projectID := tfModel.ProjectId.ValueString()
	connectionID := tfModel.Id.ValueString()
      diags := &resp.Diagnostics
	connV2 := d.Client.AtlasV2
	streamsPrivateLinkConnection, _, err := connV2.StreamsApi.GetPrivateLinkConnection(ctx, projectID, connectionID).Execute()
	if err != nil {
		diags.AddError("error fetching resource", err.Error())
		return
	}

	newStreamPrivatelinkEndpointModel := NewTFModel(ctx, diags, projectID, streamsPrivateLinkConnection)
	if diags.HasError() {
		return
	}
     diags.Append(resp.State.Set(ctx, newStreamPrivatelinkEndpointModel)...)
```

**Review by lantoli — APPROVED** [review]

**Review by kanchana-mongodb — APPROVED** [review]
